module/VuFind/src/VuFind/Db/Service/UserCardServiceInterface.php

```diff
@@ -0,0 +1,138 @@
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Sudharma Kellampalli <skellamp@villanova.edu>
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+
+namespace VuFind\Db\Service;
+
+use VuFind\Db\Entity\UserCardEntityInterface;
+use VuFind\Db\Entity\UserEntityInterface;
+
+/**
+ * Database service for UserCard.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+interface UserCardServiceInterface extends DbServiceInterface
+{
+    /**
+     * Get user_card rows with insecure catalog passwords
+     *
+     * @return UserCardEntityInterface[]
+     */
+    public function getInsecureRows(): array;
+
+    /**
+     * Get all library cards associated with the user.
+     *
+     * @param int|UserEntityInterface $user        User object or identifier
+     * @param ?int                    $id          Optional card ID filter
+     * @param ?string                 $catUsername Optional catalog username filter
+     *
+     * @return UserCardEntityInterface[]
+     */
+    public function getLibraryCards(int|UserEntityInterface $user, ?int $id = null, ?string $catUsername = null): array;
+
+    /**
+     * Get or create library card data.
+     *
+     * @param int|UserEntityInterface $user User object or identifier
+     * @param ?int                    $id   Card ID to fetch (or null to create a new card)
+     *
+     * @return UserCardEntityInterface Card data if found, null otherwise
+     * @throws \VuFind\Exception\LibraryCard
+     */
+    public function getOrCreateLibraryCard($user, $id = null): ?UserCardEntityInterface;
+
+    /**
+     * Delete library card
+     *
+     * @param UserEntityInterface     $user     User owning card to delete
+     * @param UserCardEntityInterface $userCard UserCard id or object to be deleted
+     *
+     * @return bool
+     * @throws \Exception
+     */
+    public function deleteLibraryCard(UserEntityInterface $user, int|UserCardEntityInterface $userCard): bool;
+
+    /**
+     * Save library card with the given information
+     *
+     * @param int|UserEntityInterface $user     User object or identifier
+     * @param ?int                    $id       Card ID (null = create new)
+     * @param string                  $cardName Card name
+     * @param string                  $username Username
+     * @param string                  $password Password
+     * @param string                  $homeLib  Home Library
+     *
+     * @return UserCardEntityInterface
+     * @throws \VuFind\Exception\LibraryCard
+     */
+    public function saveLibraryCard(
```

Comment:
Regardless of naming, I agree that we should support entity interfaces as much as possible, so I've adjusted the card parameter to accept IDs or objects.

Reply:
I think the most important thing would be to somehow make clear the dual functionality of creating and updating. Which `updateLibraryCardEntry()` seems to have too, actually? But obviously something like `createOrUpdateAndPersistLibraryCardEntryFromUser()` is maybe a bit too much. 😄 This could of course indicate that maybe the method itself is doing too much, but indeed no point to refactor things like that in this PR.

In any case there is always the method DocBlock for detailed documentation, and I am happy with your suggestions, or as is.

Reply:
Thanks, @aleksip -- I refined method names and comments in [aa2ee7b](https://github.com/vufind-org/vufind/pull/3679/commits/aa2ee7b8f0d5a3bffdaf91f19d68105fdaab0205); hopefully that's an improvement at least!

Reply:
In any case, I ran out of time before I could merge this today (since I want to allow time to reconcile changes with #2233 after the merge). I'll finish tomorrow, but if you have any suggested improvements in the meantime, it's not too late!

Reply:
Looks good!